README
# Asynchronous Open URI

  A CommonJS module inspired by Rubys Open-URI library.


## Install

  It's available on npm, so a simple `npm install open-uri` should be enough.


## Usage

	var open = require("open-uri");

	// Get a website
	open("http://google.com",function(err,google){console.log(google)})

	// Get a file
	open("/var/log/system.log",function(err,log){console.log(log)})

	// Stream a file to STDOUT
	open("file:///var/log/system.log",process.stdout)

	// Stream a website to a file
	var file = require("fs").createWriteStream("/tmp/goog.html")
	open("https://encrypted.google.com/search?q=open+uri",file)

	// Chain it 
	open("http://google.com",console.log)("http://publicclass.se",process.stdout)


## Supported schemes

  http, https & file


### Options for all schemes

* `stream`    (boolean) For when you only want the callback to receive a stream instead of the complete body as the second argument. Defaults to false.


### Options for HTTP(S)

* `follow`    (boolean) Follow redirects. Defaults to true.

* `headers`   (object)  Headers to pass along with the HTTP request.


## TODO

*  _HTTP(S)_ Handle different Content-Types: parse JSON if it's application/json (also handle JSONP?), return a buffer etc.

*  _HTTP(S)_ Possibly support older versions of Node

*  _HTTP(S)_ Proxy support?

* Get the tests to pass, having some issues with faking a Writeable Stream.


## Thanks to

* [Mikeal Rogers](https://github.com/mikeal), for his excellent [request module](https://github.com/mikeal/request/).

* [TJ Holowaychuck](https://github.com/visionmedia), for his minimalist projects inspiring this little tool.


## License 

(The MIT License)

Copyright (c) 2011 Robert Sk&ouml;ld &lt;robert@publicclass.se&gt;

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

History.md
# 0.4.3 / 2012-08-13

* [Change] Now using built-in zlib for gzip support (thanks to [sphericalhorse](https://github.com/sphericalhorse))

# 0.4.2 / 2012-08-13

* [Change] Rewrote tests to use [mocha](https://github.com/visionmedia/mocha) and [should](https://github.com/visionmedia/should.js).
* [Change] Compatibility with Node 0.8 (thanks to [wdavidw](https://github.com/wdavidw))
* [Change] Fully asynchronous (thanks to [wdavidw](https://github.com/wdavidw))

# 0.4.1 / 2012-03-02

* [Fix] Added `buffer`-option to file and ftp schemes.
* [Change] Updated mime dependency to 1.2.5.

# 0.4.0 / 2012-03-02

* [Change] Added the `buffer` option, which is an inverted `stream`, which is now deprecated. 
* [Fix] Avoid a global variable (by [senorpedro](https://github.com/senorpedro/))
* [Fix] _HTTP_/_HTTPS_ Only 301, 302 and 307 is proper statusCodes for redirects.

# 0.3.5 / 2011-07-15

* [Fix] Use Buffer.byteLength(body) only on strings, and body.length on Buffers.
* [Fix] Support for broken redirect implementations that give relative Location: headers.

# 0.3.4 / 2011-07-11

* [Fix] Use Buffer.byteLength(body) for Content-Length header as well, it's just better (and fixes a bug where the body was cut off).

# 0.3.3 / 2011-07-05

* [Fix] Avoid 'Out of bounds'-errors by using Buffer.byteLength(chunk) instead of chunk.length.

# 0.3.2 / 2011-05-22

* [Fix] Errors was being thrown when using userinfo in HTTP uris.

# 0.3.1 / 2011-05-22

* [Fix] Updated `addressable` dependency to 0.3.3. Now both addressable.URI and Nodes built-in URL objects are valid as the `uri` argument.
* [Fix] Fixed error thrown when an error occurs in utils.buffer().

# 0.3.0 / 2011-05-01

* [Feature] Updated `addressable` dependency to 0.3.1.

* [Feature] _HTTPS_ Added key and certificate parsing.

* [Feature] _HTTP_/_HTTPS_ Added a `method` option can be used if anything other than GET is desired.

* [Feature] _HTTP_/_HTTPS_ Added a `body` options for passing a payload with the request. Can be either a ReadableStream, a String or a Buffer.

# 0.2.2 / 2011-04-26

* [Feature] Re-factored the schemes into separate files. Now adding more schemes will be much easier.

* [Feature] Added a simple binary. Usage: `open-uri http://google.com`

* [Fix] _FTP_ is now closing properly when completed or failed.

# 0.2.1 / 2011-04-12

* [Feature] Better parsing of content using the [mime module](https://github.com/bentomas/node-mime) for Content-Type lookup.

* [Feature] _FTP_ Initial support using the [node-ftp module](https://github.com/mscdex/node-ftp).


# 0.2.0 / 2011-02-16

* [Feature] _HTTP(S)_ Now supports gzip responses (and adds a 'Accept-Encoding: gzip'-header if there is none already).

* [Feature] _HTTP(S)_ Now decodes JSON if response is of 'Content-Type: application/json'.

* [Fix] _HTTP(S)_ Fixed an issue with redirects.


# 0.1.0 / 2011-02-08

* Initial version.


test/test.streams.js
var open = require('../')

// Stream from HTTP to both FTP and File at once...
var xyz = open('https://encrypted.google.com/search?q=open+uri')
// xyz.pipe(open('ftp://abc'))
xyz.pipe(open('file:///tmp/enc-google-'+Date.now()+'.html'))


test/test-open-uri.js
var open = require("../lib/open-uri")
  , addressable = require("addressable")
  , assert = require("assert")
  , Stream = require("stream").Stream
  , port = 65000;



function writeStream(){
  var stream = new Stream();
  stream.written = false;
  stream.ended = false;
  stream.writable = true;
  stream.write = function(d){stream.written = true};
  stream.end = function(d){stream.ended = true};
  return stream;
}

// A simple test echo server
function echo(port,fn){
  var http = require("http").createServer(function(req,res){
    var headers = {'Content-Type': req.headers["content-type"] || 'text/plain'}
    if( req.headers["content-length"] ) 
      headers["Content-Length"] = req.headers["content-length"];
    res.writeHead(200, headers)
    req.pipe(res)
  })
  http.listen(port,function(){fn(http)})
}

exports["GET a website"] = function(beforeExit){
  var loaded = false;
  open("http://google.com",function(err,google){
    loaded = true;
    assert.ifError(err)
    assert.type(google,"string")
    assert.ok(google.length>0)
  })
  beforeExit(function(){assert.ok(loaded)})
}

exports["GET a website with an addressable.URI object"] = function(beforeExit){
  var loaded = false;
  var url = addressable.parse("http://google.com");
  open(url,function(err,google){
    loaded = true;
    assert.ifError(err)
    assert.type(google,"string")
    assert.ok(google.length>0)
  })
  beforeExit(function(){assert.ok(loaded)})
}

exports["GET a website with a node.js built-in URL object"] = function(beforeExit){
  var loaded = false;
  var url = require("url").parse("http://google.com");
  open(url,function(err,google){
    loaded = true;
    assert.ifError(err)
    assert.type(google,"string")
    assert.ok(google.length>0)
  })
  beforeExit(function(){assert.ok(loaded)})
}

exports["GET an encrypted website"] = function(beforeExit){
  var loaded = false;
  open("https://github.com",function(err,github){
    loaded = true;
    assert.ifError(err)
    assert.type(github,"string")
    assert.ok(github.length>0)
  })
  beforeExit(function(){assert.ok(loaded)})
}

exports["GET a website with auth"] = function(beforeExit){
  var loaded = false;
  open("http://user:pass@google.com",function(err,google){
    loaded = true;
    assert.ifError(err)
    assert.type(google,"string")
    assert.ok(google.length>0)
  })
  beforeExit(function(){assert.ok(loaded)})
}

exports["GET an encrypted website with auth"] = function(beforeExit){
  var loaded = false;
  open("https://user:pass@google.com",function(err,google){
    loaded = true;
    assert.ifError(err)
    assert.type(google,"string")
    assert.ok(google.length>0)
  })
  beforeExit(function(){assert.ok(loaded)})
}

exports["POST a string to a website"] = function(){
  echo(++port,function(server){
    open("http://localhost:"+port,{method:"POST",body:"abc"},function(err,dump,res){
      server.close()
      assert.ifError(err)
      assert.equal(res.headers["content-type"],"text/plain")
      assert.equal(dump.toString(),"abc")
    })
  })
}

exports["POST a buffer to a website"] = function(){
  echo(++port,function(server){
    open("http://localhost:"+port,{method:"POST",body:new Buffer([1,2,3,4]),headers:{"Content-Type":"application/octet-stream"}},function(err,dump,res){
      server.close()
      assert.ifError(err)
      assert.equal(res.headers["content-type"],"application/octet-stream")
      assert.ok(Buffer.isBuffer(dump))
      assert.equal(dump.length,4)
    })
  })
}

exports["PUT a form to a website"] = function(){
  echo(++port,function(server){
    open("http://localhost:"+port,{method:"POST",body:"a=1&b=2&c=3",headers:{"Content-Type":"application/x-www-form-urlencoded"}},function(err,dump,res){
      server.close()
      assert.ifError(err)
      assert.equal(res.headers["content-type"],"application/x-www-form-urlencoded")
      assert.type(dump,"object")
      assert.eql(dump,{a:1,b:2,c:3})
    })
  })
}

exports["POST some json to a website"] = function(){
  echo(++port,function(server){
    open("http://localhost:"+port,{method:"POST",body:'{"a":1,"b":2,"c":3}',headers:{"Content-Type":"application/json"}},function(err,dump,res){
      server.close()
      assert.ifError(err)
      assert.equal(res.headers["content-type"],"application/json")
      assert.type(dump,"object")
      assert.eql(dump,{a:1,b:2,c:3})
    })
  })
}


exports["POST stream text to a website"] = function(){
  echo(++port,function(server){
    var file = require("fs").createReadStream("README.md");
    open("http://localhost:"+port,{method:"POST",body:file},function(err,dump,res){
      server.close()
      assert.ifError(err)
      assert.equal(res.headers["content-type"],"text/plain")
      assert.type(dump,"string")
      assert.eql(dump,require("fs").readFileSync("README.md","utf8"))
    })
  })
}

exports["GET a relative file"] = function(beforeExit){
  var loaded = false;
  open("README.md",function(err,log){  
    loaded = true;
    assert.ifError(err)
    assert.ok(Buffer.isBuffer(log))
    assert.type(log.toString(),"string")
    assert.ok(log.length>0)
  })
  beforeExit(function(){assert.ok(loaded)})
}

exports["GET Stream an absolute file"] = function(beforeExit){
  var stream = writeStream();
  open("file:///var/log/system.log",stream)
  beforeExit(function(){
    assert.ok(stream.written)
    assert.ok(stream.ended)
  })
}
 
exports["GET Stream a website to a file"] = function(beforeExit){
  var path = "/tmp/goog-"+Date.now()+".html";
  var file = require("fs").createWriteStream(path)
  open("https://encrypted.google.com/search?q=open+uri",file)
  beforeExit(function(){
    assert.doesNotThrow(function(){
      require("fs").statSync(path)
    })
  })
}

exports["GET a text file from ftp"] = function(beforeExit){
  var loaded = false;
  open("ftp://ftp.sunet.se/pub/Internet-documents/rfc/rfc100.txt",function(err,rfc){
    loaded = true
    assert.ifError(err)
    assert.type(rfc,"string")
    assert.ok(rfc.length > 0)
  })
  beforeExit(function(){assert.ok(loaded)})
}

exports["Attempt to get a non-existing text file from ftp"] = function(beforeExit){
  var loaded = false;
  open("ftp://ftp.sunet.se/im-not-here.txt",function(err,rfc){
    loaded = true
    assert.ok(err)
    assert.type(rfc,"undefined")
  })
  beforeExit(function(){assert.ok(loaded)})
}

exports["Stream a text file from ftp to a file"] = function(beforeExit){
  var path = "/tmp/rfc-"+Date.now()+".html";
  var file = require("fs").createWriteStream(path)
  open("ftp://ftp.sunet.se/pub/Internet-documents/rfc/rfc100.txt",file)
  beforeExit(function(){
    assert.doesNotThrow(function(){
      require("fs").statSync(path)
    })
  })
}

exports["GET a redirect with a relative Location"] = function(beforeExit){
  var loaded = false;
  open("http://golang.org/cmd/5a",function(err,go,res){
    loaded = true
    assert.ifError(err)
    assert.includes(go,'<title>Command 5a - The Go Programming Language</title>')
  })
  beforeExit(function(){assert.ok(loaded)})
}

exports["GET a redirect with a relative Location without 'follow'"] = function(beforeExit){
  var loaded = false;
  open("http://golang.org/cmd/5a",{follow:false},function(err,go,res){
    loaded = true
    assert.ifError(err)
    assert.equal(res.statusCode,301)
    assert.includes(go,'Moved Permanently')
  })
  beforeExit(function(){assert.ok(loaded)})
}

exports["Chain it"] = function(beforeExit){
  var stream = writeStream();
  var loaded = false;
  open("http://google.com",stream)("file:///var/log/system.log",function(err,google){
    loaded = true;
    assert.type(google.toString(),"string")
    assert.ok(google.length>0)
  })
  beforeExit(function(){
    assert.ok(loaded)
    assert.ok(stream.written)
    assert.ok(stream.ended)
  })
}

exports["Throw when error without callback."] = function(){
  assert.throws(function(){
    open("/i-dont-exist!")
  })
}

exports["Does not throw when error with a callback."] = function(){
  assert.doesNotThrow(function(){
    open("/i-dont-exist!",function(err){ assert.ok(err) })
  })
}

test/test.open-uri.js
var open = require('../');

describe('open-uri',function(){

  it('should throw when error and no callback',function(){
    (function(){
      open('/i-dont-exist!')
    }).should.throw
  })

  it('should not throw when error and callback',function(){
    (function(){
      open('/i-dont-exist!',function(err){
        err.should.exist
      })
    }).should.not.throw
  })
})


lib/utils.js
var mime = require("mime")
  , open = require("./open-uri");

/**
* Helper for buffering up a stream and passing its content to a callback.
* 
* @param {String} type
*     The mime type of the stream.
* @param {String} path
*     The path to lookup the type from in case it's not available.
* @param {ReadableStream} stream
*     A ReadableStream to buffer from.
* @param {Function} output
*     Callback which is called whenever the buffering is completed or if there was an error.
*/
exports.buffer = function buffer(type,path,stream,output){
  // Fix the type if it comes from a HTTP header
  type = (type||'').split(";")[0];
  var mimeType = mime.lookup(type||path||'')
    , encoding = mime.charsets.lookup(mimeType, 'utf8')
    , buf = []
    , len = 0;
  stream.on("error",function(err){ 
    // TODO Close the stream on error?
    error(output, err); 
  }) 
  stream.on("timeout",function(){ 
    // TODO Close the stream on timeout?
    error(output,new Error("[OpenURI] Connection timed out.")); 
  }) 
  stream.on("data",function(chunk){
    chunk = chunk.toString(encoding);
    buf.push(chunk);
    len += Buffer.byteLength(chunk);
  })
  stream.on("end",function(){
    try {
      var data = parse(type,buf,len) 
              || parse(mimeType,buf,len) 
              || parse.default(buf,len);
      output(null,data,stream);
    } catch(e){
      error(output,e)
    }
  })
}

/**
* Helper for parsing the content based on mime-type. Ex. parsing 
* a JSON string to an object, text based content to a string and 
* binary as a buffer.
* 
* @param {String} type
*     The mime-type.
* @param {Array<Buffer>} buf
*     An array of Buffer objects.
* @param {Integer} len
*     The total size of the buffers.
* @return The parsed content.
*/
function parse(type,buf,len){
  for( var key in parse )
    if( ~type.indexOf(key) )
      return parse[key](buf,len);
  return false;
}
parse["text/"] = function(buf,len){ 
  return buf.join("") 
}
parse["application/json"] = function(buf,len){ 
  return JSON.parse(buf.join("")) 
}
parse["application/x-www-form-urlencoded"] = function(buf,len){ 
  return require("querystring").parse(buf.join("")) 
}
parse.default = 
parse["application/octet-stream"] = function(buf,len){
  var offset = 0;
  return buf.reduce(function(body,b){
    offset += body.write(b,offset)
    return body;
  },new Buffer(len))
}
exports.parse = parse;

/**
* Helper for returning or throwing errors depending on the type 
* of `output`.
* 
* @param {Function|WritableStream|Null} output
*     The output callback or stream.
* @param {Error} err
*     The Error to throw/call.
* @returns `open` for chainability.
*/
function error(output,err){
  if( typeof output == "function" )
    output(err);
  else 
    throw( err );
  return open;
}
exports.error = error;

/**
* Converts string to base64.
*
* @param {String|Buffer|Array} str
*     Buffer friendly object to convert.
* @return a Base64 encoded string.
*/
exports.toBase64 = function toBase64(str){ 
  return (new Buffer(str||"", "ascii")).toString("base64") 
}

/**
* Checks if the output is a valid callback or stream.
* 
* @param {String|WritableStream} output
*     The output to check against.
* @return `true` if it's valid, otherwise `false`.
*/
exports.isValidOutput = function isValidOutput(output){
  return typeof output == "function" || ( typeof output == "object" && output.writable && "pipe" in output )
}


lib/open-uri.js
/**
*  Asynchronous Open URI, inspired by Rubys Open-URI library.
*  
*  @param   {URI|String} uri                    The URI to open.
*  @param   {Object} opts                       (optional) Options for the scheme (see each scheme for details). 
*  @param   {Function|WriteableStream}  output  Where to send the data.
*  @return  Itself for chainability.
*/
var addressable = require("addressable")
  , utils = require("./utils")
  , inspect = require("util").inspect;

module.exports = function open(uri,opts,output){
  if( utils.isValidOutput( opts ) )
    output = opts, opts = {};
    
  if( !utils.isValidOutput( output ) )
    return utils.error( output, new Error("[OpenURI] Invalid output: " + inspect(arguments)) );
  
  if( !(uri = addressable.parse(uri)) ) 
    return utils.error( output, new Error("[OpenURI] Invalid URI: " + inspect(arguments[0])) );
  
  try {
    require("./schemes/"+(uri.scheme||"file")).call(this,uri,opts,output);
  } catch( e ){
    console.error(e.stack)
    return utils.error( output, new Error("[OpenURI] Invalid scheme: " + inspect(uri)) );
  }
  
  return open;
}

lib/schemes/ftp.js
/**
*   FTP Scheme
*
*   Available options:
*
*     {Boolean} buffer    For when you only want the body to be buffered and returned as the second argument instead of receiving the direct stream. Defaults to `true`.
*     {Boolean} anonymous If no user info is in the URI, add anonymous. Defaults to `true`.
*/
var ftp = require("ftp")
  , utils = require("../utils");

module.exports = function(uri,opts,output){
  var client = new ftp()
    , authenticated = false;

  // Default to buffer (using `opts.stream` for backward compability)
  if( !opts.hasOwnProperty('buffer') )
    opts.buffer = !opts.stream;
  
  if( !uri.userinfo && opts.anonymous !== false ){
    uri.username = "anonymous";
    uri.password = "anonymous@";
  }
  
  client.on("connect",function request(err){
    if( err ) return utils.error(output,err);
    
    // Authenticate first if uri has userinfo or opts.anonymous is set.
    if( uri.userinfo && !authenticated ){
      return client.auth(uri.username,uri.password,function(err){
        authenticated = !err
        request(err)
      })
    }
    
    // TODO Add this to addressable? or maybe it's in the path-module?
    var isDir = uri.path.charAt(uri.path.length-1) == "/";
    if( isDir ){
      // TODO If path is a directory, we client.list(uri.pathname)
      utils.error( output, new Error("Not implemented yet, listing a directory with ftp://"));
    } else if( opts.body ){
      // TODO If the path is a file, and body is in opts, we upload it with client.put(uri.pathname)
      utils.error( output, new Error("Not implemented yet, uploading a file to ftp://"));
    } else {
      // If the path is a file, client.get(uri.path) and do the stream magic we do in open.file()
      client.get(uri.pathname,function(err,stream){
        if( err ) return utils.error( output, err );
        // Make sure the client closes when done/failed.
        stream.on("error",function(){ client.end() })
        stream.on("end",function(){ client.end() })
        if( typeof output == "function" ){
          if( opts.buffer ){
            utils.buffer(null,uri.pathname,stream,output)
          } else {
            output(null,stream);
          }
        } else {
          stream.pipe(output);
        }
      })
    }
  }).connect(uri.port,uri.host)
}

lib/schemes/file.js
/**
*   File Scheme
*
*   Available options:
*
*     {Boolean} buffer    For when you only want the body to be buffered and returned as the second argument instead of receiving the direct stream. Defaults to `true`.
*     {String}  encoding  The encoding of the file to read.
*/
var fs = require("fs")
  , path = require("path")
  , utils = require("../utils")
  , exists = fs.exists || path.exists;

module.exports = function(uri,opts,output){
  // Default to buffer (using `opts.stream` for backward compability)
  if( !opts.hasOwnProperty('buffer') )
    opts.buffer = !opts.stream;

  if( uri.isRelative() ){
    uri.pathname = path.join(process.cwd(),uri.pathname);
    uri.normalize()
  }

  exists(uri.pathname, function(exists){
    if( !exists )
      return utils.error(output,new Error("[OpenURI] File Not Found: "+uri.pathname))

    var args = opts.start && opts.end 
      ? [uri.pathname,opts] // with 'range'
      : [uri.pathname];     // no 'range'

    try {
      if( typeof output == "function" ){
        if( opts.buffer ){
          fs.readFile(uri.pathname,opts.encoding,output);
        } else {
          output(null,fs.createReadStream.apply(fs,args));
        }
      } else {
        fs.createReadStream.apply(fs,args).pipe(output)
      }
    } catch(e) {
      utils.error(output,e);
    }
  })
}


lib/schemes/https.js
/**
*   HTTPS Scheme
*
*   Available options:
*
*     {Boolean} stream    For when you only want the callback to receive a stream instead of the complete body as the second argument. Defaults to `false`.
*     {Boolean} follow    Follow redirects. Defaults to `true`.
*     {Object}  headers   Headers to pass along with the HTTP request.
*     {Boolean} gzip      If the request should be attempted with gzip. Defaults to `true` if the 'node-compress'-module is available.
*     {String}  method    HTTP request method. Default 'GET'.
*     {String}  key       Private key to use for SSL. Defaults to ENV var NODE_HTTPS_KEY or attempts to find it as "./key.pem".
*     {String}  cert      Public x509 certificate to use. Defaults to ENV var NODE_HTTPS_CERT or attempts to find it as "./cert.pem".
*     {String|Array} ca   An authority certificate or array of authority certificates to check the remote host against.
*/
var addressable = require("addressable")
  , fs = require("fs")
  , path = require("path")
  , http = require("./http")
  , exists = fs.exists || path.exists
  , read = fs.readFile;

module.exports = function https(uri,opts,output){
  // Looks for SSL Key and Certificate in:
  //   1. In the options as a normal https.get(). Should it expand them if it's a path? I.e. {key: "./key.pem", cert: "./cert.pem"} is fs.readSync()-ed?
  //   2. ENV-variables to their paths, NODE_HTTPS_KEY=./key.pem NODE_HTTPS_CERT=./cert.pem
  //   3. Check for them in the process pwd.
  
  var key  = opts.key  || process.env.NODE_HTTPS_KEY  || "./key.pem"
  var cert = opts.cert || process.env.NODE_HTTPS_CERT || "./cert.pem"
  var pending = 2;

  // Call this when key/cert is loaded
  var download = http.bind(this,uri,opts,output);
  
  // Load Key if it's a file
  readIfFile(key,function(key){
    opts.key = key;
    --pending || download();
  })

  // Load Certificate if it's a file
  readIfFile(cert,function(cert){
    opts.cert = cert;
    --pending || download();
  })
}

function readIfFile(file,fn){
  var uri = addressable.parse( file );

  // If its a file:// or no scheme, check if it
  // exists and if so read it's contents. otherwise
  // assume it's the proper data already
  if( uri && (uri.scheme == "file" || !uri.scheme) ){
    exists(uri.pathname,function(yes){
      if( yes ) 
        read(uri.pathname,function(err,data){
          fn(data||file);
        });
      else 
        fn(file)
    })
  } else {
    fn(file)
  }
}

lib/schemes/http.js
/**
*   HTTP Scheme
*
*   Available options:
*
*     {Boolean} buffer
*       For when you only want the body to be buffered and returned as the second argument instead of receiving the direct stream. Defaults to `true`.
*     {Boolean} follow
*       Follow redirects. Defaults to `true`.
*     {Object}  headers
*       Headers to pass along with the HTTP request.
*     {Boolean} gzip
*       If the request should be attempted with gzip. Defaults to `true`.
*     {String} method
*       HTTP request method. Default 'GET'. 
*     {ReadableStream|Buffer|String} body
*       Body to write to the request.
*/

var addressable = require("addressable")
  , open = require("../open-uri")
  , mime = require("mime")
  , utils = require("../utils")
  , Buffer = require("buffer").Buffer
  , zlib = require("zlib");

module.exports = function http(uri,opts,output){
  uri.headers = opts.headers = opts.headers || {}

  // Default to gzip
  if( !opts.hasOwnProperty('gzip') )
    opts.gzip = true

  // Default to buffer (using `opts.stream` for backward compability)
  if( !opts.hasOwnProperty('buffer') )
    opts.buffer = !opts.stream;
  
  // HTTP Auth
  if( uri.userinfo && !uri.headers.authorization )
    uri.headers.authorization = "Basic " + utils.toBase64(uri.userinfo);
  
  // Use gzip?
  if( opts.gzip && !uri.headers["accept-encoding"] && !uri.headers["Accept-Encoding"] )
    uri.headers["accept-encoding"] = "gzip";
  
  // HTTP method
  uri.method = opts.method || "GET";
  
  // Body Content-Length
  if( typeof opts.body == "string" )
    uri.headers["content-length"] = Buffer.byteLength( opts.body );
  else if( Buffer.isBuffer( opts.body ) )
    uri.headers["content-length"] = opts.body.length;
  
  var req = require(uri.scheme).request(uri);

  // Handle response
  req.on('response',function(res){

    // Follow Redirects
    if( opts.follow !== false && ~[301,302,307].indexOf(res.statusCode) && res.headers.location ){
      var location = res.headers.location;
      // It's an invalid 'relative' Location, prepend with current host
      if( location[0] == '/' ) 
        location = uri.scheme + '://' + uri.authority + location;
      opts.headers.host = addressable.parse(location).host;
      return open(location,opts,output);
    }
    
    // Handle gzip responses
    var stream = res;
    if( opts.gzip && ~(res.headers['content-encoding']||"").indexOf("gzip") )
      res.pipe(stream = zlib.createUnzip());
    
    if( typeof output == "function" ){
      if( opts.buffer ){
        utils.buffer(res.headers["content-type"],uri.pathname,stream,function(err,body){
          output(null,body,res);
        })
      } else {
        output(null,stream);
      }
    } else {
      stream.pipe(output);
    }
  })

  // Request error
  req.on('error',function(e){
    utils.error(output,e)
  })
  
  // Handle a request body.
  if( typeof opts.body == "string" || Buffer.isBuffer( opts.body ) )
    req.end(opts.body);
  else if( typeof opts.body == "object" && "pipe" in opts.body && opts.body.readable )
    opts.body.pipe(req);
  else
    req.end();
}
